### ISSUE TITLE
Make various geogram functions thread-safe

### ISSUE THREAD

**Issue by jdumas** [BrunoLevy/geogram#68]
By "thread-safe" here I mean "being able to call geogram functions concurrently". E.g. being able to run Delaunay triangulation on two meshes in parallel (which can happen when integrating Geogram functions in a node-base application like Blender). This is why having a central scheduler (like OneTBB) helps a lot. Currently a lot of geogram routines use global static variables. Using `thread_local` instead would help. Some examples:
- [CVT.h](https://github.com/BrunoLevy/geogram/blob/ec0a77e455435f365b7b9d5577dfbc6f732350a2/src/lib/geogram/voronoi/CVT.h#L442)
- [lbfgs_optimizers.cpp](https://github.com/BrunoLevy/geogram/blob/ec0a77e455435f365b7b9d5577dfbc6f732350a2/src/lib/geogram/numerics/lbfgs_optimizers.cpp#L66)
- [HLBFGS/LineSearch.cpp](https://github.com/BrunoLevy/geogram/blob/ec0a77e455435f365b7b9d5577dfbc6f732350a2/src/lib/geogram/third_party/HLBFGS/LineSearch.cpp#L17-L27)
- [HLBFGS/ICFS.cpp](https://github.com/BrunoLevy/geogram/blob/ec0a77e455435f365b7b9d5577dfbc6f732350a2/src/lib/geogram/third_party/HLBFGS/ICFS.cpp#L50-L52)

Ideally this should be tested with `ThreadSanitizer` by calling Delaunay triangulation on a few meshes in parallel...

**Comment by BrunoLevy** [comment]
ForLBFGS/ HLBFGS, Dmitry Sokolov wrote a new version, that probably does not have all these globals. I'll take a look !

**Comment by Abenabbou** [comment]
Hi Bruno,
We have the same issue in SLB for 2D Delaunay triangulation. We cannot call this algorithm in parallel. Actually, we can have millions of 2d patchs (point sets) that we want to triangulate in parallel. We use openMP (we tried also with TBB) to parallelize the call to the 2D Delaunay triangulation.

Please let me know if you need data set to reproduce the issue.

Thanks,
A. Benabbou

**Comment by BrunoLevy** [comment]
Hi Azeddine,

Yes, the `CDT2d` class has the option to restore Delaunayness after constraints insertion, and does not insert Steiner points. It also handles intersecting constraints (then it automatically inserts the intersection in the triangulation)

**Comment by Abenabbou** [comment]
Thank you Bruno. We will try it.

**Comment by jdumas** [comment]
> ForLBFGS/ HLBFGS, Dmitry Sokolov wrote a new version, that probably does not have all these globals. I'll take a look !

Is this version available somewhere btw?

**Comment by BrunoLevy** [comment]
Yes, [here](https://github.com/ssloy/stlbfgs)
(Note: licensed under Afferro GPL)

**Comment by jdumas** [comment]
Would you be able to use it in Geogram under a different license then?

**Comment by jdumas** [comment]
If Dmitry doesn't want to release his implementation under a permissive license, it'd be good to make it possible to plug other L-BFGS implementations into Geogram, e.g. via the use of a templated or polymorphic interface. E.g. I've used [CppNumericalSolvers](https://github.com/PatWie/CppNumericalSolvers) in the past and I like how it's designed. There's also [optim](https://github.com/kthohr/optim) which has a permissive license and a few interesting features. In any case it'd be good to make the L-BFGS backend solver configurable so we can hook up a thread-safe implementation for industrial applications.

**Comment by BrunoLevy** [comment]
There is also [LBFGS++](https://lbfgspp.statr.me/), will take a look at it (MIT license)

**Comment by BrunoLevy** [comment]
Cedric has sent to me a patch for making the ProgressLogger thread safe (to be included in Sept or Oct)

**Comment by jdumas** [comment]
> There is also [LBFGS++](https://lbfgspp.statr.me/), will take a look at it (MIT license)

Yeah we're using LBFGS++ internally and it works great. The only downside for you is that it depends on Eigen!
